test/url.cpp

```diff
@@ -854,6 +854,122 @@ class url_test
             it--;
             BOOST_TEST(it->encoded_key() == "x");
         }
+
+        {
+            url u("http://user:pass@example.com:80/file.txt");
+            auto qp = u.params();
+            BOOST_TEST(qp.empty());
+            qp.insert_encoded(
+                qp.insert_encoded(
+                    qp.insert_encoded(
+                        qp.insert_encoded(
+                            qp.insert_encoded(qp.end(), {"e", "5"} ), {"d", "4"} ), {"c", "3"} ), {"b", "2"} ), {"a", "1"} );
+            BOOST_TEST(!qp.empty());
+            BOOST_TEST(qp.size() == 5);
+            BOOST_TEST(std::distance(qp.begin(), qp.end()) == 5);
+            BOOST_TEST(u.encoded_query() == "a=1&b=2&c=3&d=4&e=5");
+
+            {
+                auto e = qp.begin();
+                ++e;
+                ++e;
+                qp.erase( qp.begin(), e );
+            }
+            BOOST_TEST(!qp.empty());
+            BOOST_TEST(qp.size() == 3);
+            BOOST_TEST(std::distance(qp.begin(), qp.end()) == 3);
+            BOOST_TEST(u.encoded_query() == "c=3&d=4&e=5");
+
+            auto after = qp.insert_encoded(qp.begin(), {"a","1"});
+            ++after;
+            BOOST_TEST(!qp.empty());
+            BOOST_TEST(qp.size() == 4);
+            BOOST_TEST(std::distance(qp.begin(), qp.end()) == 4);
+            BOOST_TEST(u.encoded_query() == "a=1&c=3&d=4&e=5");
```

Comment:
we need to know what happens with degenerate query strings. For example what if the query string is
`?&`
`?&&`
`?=`
`?=&`
`?&=`
`?&=&`

And you try to insert?